text/0000-must-not-await-lint.md

```diff
@@ -0,0 +1,152 @@
+
+- The same applies to other such synchronization primitives such as locks from `parking-lot`.
+
+- `tracing::Span` has the ability to enter the span via the `tracing::span::Entered` guard. While entering a span is totally normal, during an async fn the span only needs to be entered once before the `.await` call, which might potentially yield the execution.
+
+- Any RAII guard might possibly create unintended behavior if held across an await boundary.
+
+This lint will enable the compiler to warn the user that the code could produce unforeseen side effects. Some examples of this are:
+
+- [`std::sync::MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)
+- [`tracing::span::Entered`](https://docs.rs/tracing/0.1.15/tracing/span/struct.Entered.html)
+
+This will be a best effort lint to signal the user about unintended side-effects of using certain types across an await boundary.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `must_not_await` attribute is used to issue a diagnostic warning when a value is not "used". It can be applied to user-defined composite types (structs, enums and unions), functions and traits.
+
+The `must_not_await` attribute may include a message by using the [`MetaNameValueStr`] syntax such as `#[must_not_await = "example message"]`.  The message will be given alongside the warning.
+
+When used on a user-defined composite type, if a value exists across an await point, then this lint is violated.
+
+
+```rust
+#[must_not_await = "Your error message here"]
+struct MyStruct {}
+
+async fn foo() {
+  let my_struct = MyStruct {};
+  my_async_op.await;
+  println!("{:?}", my_struct);
+}
+```
+
+When used on a function, if the value returned by a function is held across an await point, this lint is violated.
+
+```rust
+#[must_not_await]
+fn foo() -> i32 { 5i32 }
+
+async fn foo() {
+  let bar = foo();
+  my_async_op.await;
+  println!("{:?}", bar);
+}
+```
+
+When used on a [trait declaration], if the value implementing that trait is held across an await point, the lint is violated.
+
+```rust
+trait Trait {
+    #[must_not_await]
+    fn foo(&self) -> i32;
+}
+
+impl Trait for i32 {
+    fn foo(&self) -> i32 { 0i32 }
+}
+
+async fn foo() {
+  let bar = 5i32.foo();
+  my_async_op.await;
+  println!("{:?}", bar);
+}
+```
+
+When used on a function in a trait implementation, the attribute does nothing.
+
+[`MetaNameValueStr`]: https://doc.rust-lang.org/reference/attributes.html#meta-item-attribute-syntax
+[trait declaration]: https://doc.rust-lang.org/reference/items/traits.html
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- There is a possibility it can produce a false positive warning and it could get noisy. But using the `allow` attribute would work similar to other `deny-by-default` lints.
```

Comment:
Given that earlier this is described as for types which "might possibly create unintended behavior" (which implies to me that it's not necessarily wrong to hold them over an await), could you maybe talk about the `allow` more?  It seems like it'd be more needed than for other things -- `must_use` one can quiet with `let _ =`, for example, rather than needing `#[allow]`.

For example, at what scope can the allow be applied?  Can it be quieted for particular awaits?  Or for only particular bindings (so one still gets warning for other bindings)?

Reply:
I can certainly expand on it, but I think the behavior for `#[allow]` would be what is specified here: https://doc.rust-lang.org/reference/attributes/diagnostics.html#lint-check-attributes

I agree, that talking about how the only way to silence the lint is through the `#[allow]` attribute is important, due to the reasons you mentioned.